satellite/satellitedb/mutex.go

```diff
@@ -0,0 +1,267 @@
+	return db.statdb.FindInvalidNodes(ctx, nodeIDs, maxStats)
+}
+
+// Update all parts of single storagenode's stats in the db
+func (db *muStatDB) Update(ctx context.Context, updateReq *statdb.UpdateRequest) (stats *statdb.NodeStats, err error) {
+	defer db.mu.locked()()
+	return db.statdb.Update(ctx, updateReq)
+}
+
+// UpdateUptime updates a single storagenode's uptime stats in the db
+func (db *muStatDB) UpdateUptime(ctx context.Context, nodeID storj.NodeID, isUp bool) (stats *statdb.NodeStats, err error) {
+	defer db.mu.locked()()
+	return db.statdb.UpdateUptime(ctx, nodeID, isUp)
+}
+
+// UpdateAuditSuccess updates a single storagenode's audit stats in the db
+func (db *muStatDB) UpdateAuditSuccess(ctx context.Context, nodeID storj.NodeID, auditSuccess bool) (stats *statdb.NodeStats, err error) {
+	defer db.mu.locked()()
+	return db.statdb.UpdateAuditSuccess(ctx, nodeID, auditSuccess)
+}
+
+// UpdateBatch for updating multiple farmers' stats in the db
+func (db *muStatDB) UpdateBatch(ctx context.Context, updateReqList []*statdb.UpdateRequest) (statsList []*statdb.NodeStats, failedUpdateReqs []*statdb.UpdateRequest, err error) {
+	defer db.mu.locked()()
+	return db.statdb.UpdateBatch(ctx, updateReqList)
+}
+
+// CreateEntryIfNotExists creates a statdb node entry and saves to statdb if it didn't already exist
+func (db *muStatDB) CreateEntryIfNotExists(ctx context.Context, nodeID storj.NodeID) (stats *statdb.NodeStats, err error) {
+	defer db.mu.locked()()
+	return db.statdb.CreateEntryIfNotExists(ctx, nodeID)
+}
+
+// muOverlayCache implements a mutex around overlay cache
+type muOverlayCache struct {
+	mu      *Mutex
+	overlay storage.KeyValueStore
+}
+
+// Put adds a value to store
+func (db *muOverlayCache) Put(key storage.Key, value storage.Value) error {
+	defer db.mu.locked()()
+	return db.overlay.Put(key, value)
+}
+
+// Get gets a value to store
+func (db *muOverlayCache) Get(key storage.Key) (storage.Value, error) {
+	defer db.mu.locked()()
+	return db.overlay.Get(key)
+}
+
+// GetAll gets all values from the store
+func (db *muOverlayCache) GetAll(keys storage.Keys) (storage.Values, error) {
+	defer db.mu.locked()()
+	return db.overlay.GetAll(keys)
+}
+
+// Delete deletes key and the value
+func (db *muOverlayCache) Delete(key storage.Key) error {
+	defer db.mu.locked()()
+	return db.overlay.Delete(key)
+}
+
+// List lists all keys starting from start and upto limit items
+func (db *muOverlayCache) List(start storage.Key, limit int) (keys storage.Keys, err error) {
+	defer db.mu.locked()()
+	return db.overlay.List(start, limit)
+}
+
+// ReverseList lists all keys in revers order
+func (db *muOverlayCache) ReverseList(start storage.Key, limit int) (storage.Keys, error) {
+	defer db.mu.locked()()
+	return db.overlay.ReverseList(start, limit)
+}
+
+// Iterate iterates over items based on opts
+func (db *muOverlayCache) Iterate(opts storage.IterateOptions, fn func(storage.Iterator) error) error {
+	return errors.New("not implemented")
```

Comment:
Any reason why this is different from `ReverseList` or `Close`?

Reply:
It has a function closure. Although I probably could still wrap it inside the mutex.

Reply:
No need to complicate this. I think it can stay as it is.